build-api/api.go

```diff
@@ -20,55 +25,107 @@ type DockerInfo struct {
 	m.Run()
-
 }
 
-func BuildDockerFile(p martini.Params, r render.Render, dockerInfo DockerInfo) {
-	tmp_dir := "tmp"
+func GenerateDockerFile(dockerInfo DockerInfo, buildDir string) error {
+	var err_return error
 
 	// Create a new Dockerfile template parses template definition
 	docker_tmpl, err := template.ParseFiles("templates/dockerfile.tmplt")
-	HandleErr(err)
+	logging.LogToFile(err)
+	err_return = err
 
-	// Create tmp directory
-	err = os.Mkdir(tmp_dir, 0770)
-	HandleErr(err)
+	err = os.Mkdir(buildDir, 0770)
+	logging.LogToFile(err)
+	err_return = err
 
-	// Create file
-	f, err := os.Create(tmp_dir + "/Dockerfile")
-	HandleErr(err)
+	f, err := os.Create(buildDir + "/Dockerfile")
+	logging.LogToFile(err)
+	err_return = err
 	defer f.Close()
 
 	//Apply the Dockerfile template to the docker info from the request
 	err = docker_tmpl.Execute(f, dockerInfo)
-	HandleErr(err)
+	err_return = err
+	logging.LogToFile(err)
 
-	// Build the docker container.
-	go buildDockerContainer()
-	r.JSON(200, map[string]interface{}{"success": true})
+	return err_return
+}
+
+// grabBuiltStaticFiles issues a `docker run` command to the container image
+// we created that will transfer built files to specified location
+func GrabBuiltStaticFiles(dockerImageID, projectName, transferLocation string) {
+	// Not sure if this is the best way to handle "dynamic strings"
+	mountStringSlice := []string{transferLocation, ":", "/tmp_mount"}
+	mountString := strings.Join(mountStringSlice, "")
+
+	err := os.Mkdir(transferLocation, 0770)
+	logging.LogToFile(err)
+
+	transfer := exec.Command("scripts/transfer_files.sh", mountString, dockerImageID, projectName)
+	res, err := transfer.CombinedOutput()
+	logging.LogToFile(err)
+	logging.LogToFile(string(res))
+}
 
+func Build(buildDir, projectName, remotePath string) string {
+	c1 := make(chan string)
+	go BuildDockerContainer(c1)
+
+	// Looping until we get notification on the channel c1 that the build has finished
+	for {
+		select {
+		case buildTag := <-c1:
+			logging.LogToFile("Container built...transfering built files...")
+			GrabBuiltStaticFiles(buildTag, projectName, buildDir)
+			if os.Getenv("ENV") != "test" {
+				rsyncProject(buildDir, remotePath)
+			}
+			return "success"
+		}
+	}
+}
+
+func rsyncProject(buildDir, remoteLoc string) {
+	rsyncCommand := exec.Command("scripts/rsync_project.sh", buildDir, remoteLoc)
+	res, err := rsyncCommand.CombinedOutput()
+	logging.LogToFile(err)
+	logging.LogToFile(string(res))
+}
+
+func BuildDockerFile(p martini.Params, r render.Render, dockerInfo DockerInfo) {
+	buildLocation := os.Getenv("BUILD_LOCATION")
+	err := GenerateDockerFile(dockerInfo, ".")
+
+	if err != nil {
+		r.JSON(500, map[string]interface{}{"success": false})
+	}
+
+	logging.LogToFile("Dockerfile generated successfully...building container...")
+
+	go Build(buildLocation, dockerInfo.REPO_NAME, dockerInfo.PATH)
+	r.JSON(200, map[string]interface{}{"success": true})
```

Comment:
Lets not change this now as API is still expecting this response. But in the future, we should switch API to setting status based on 200, 400, 404, 500, etc.